README.md
SB-Communication
================

SB-Communication is a base communication low-level library meant to be used in other communication libraries.
It allows promise-based communication on both ends, You must write a wrapper for it for your use-case specific
stream, socket or resource though. It works in both browsers and node!

#### Example Communication Implementation

```js
'use babel'

import Communication from 'sb-communication'

const worker = new WebWorker('worker.js')

const communication = new Communication({
  listener(callback) {
    worker.onmessage = callback
  },
  send(data) {
    worker.postMessage = data
  },
})

// Event-Specific bindings
communication.on('count-pi', async function(data) {
  // ^ Handlers can return Promises or be sync
  console.log(data)
  return 'Something else'
})
```

#### API

```js
export default class Communication {
  constructor({ listener: Function, send: Function })
  on(event, callback)
  send(event, payload)
  dispose()
}
```

#### LICENSE

This project is licensed under the terms of MIT License.


CHANGELOG.md
#### 5.0.0

* Complete rewrite

#### 4.0.1

* Fix a bug about passing promises as response

#### 4.0.0

* Complete rewrite


spec/worker.js
// @flow

const Communication = require('../')

const communication = new Communication({
  listener(callback) {
    process.on('message', callback)
  },
  send(payload) {
    process.send(payload)
  },
})

communication.on('ping', function(arg) {
  return new Promise(function(resolve) {
    resolve(`pong ${arg}`)
  })
})
communication.on('ping2', function(arg) {
  return `pong2 ${arg}`
})
communication.on('ping3', function(arg) {
  return new Promise(function(resolve, reject) {
    reject(new Error(`pong3 ${arg}`))
  })
})
communication.on('ping4', function(arg) {
  throw new Error(`pong4 ${arg}`)
})


spec/main-spec.js
'use babel'

import path from 'path'
import { it } from 'jasmine-fix'
import childProcess from 'child_process'

import Communication from '../'

describe('SB-Communication', function() {
  let forkedProcess = null

  beforeEach(function() {
    forkedProcess = childProcess.fork(path.join(__dirname, 'worker.js'))
  })
  afterEach(function() {
    if (forkedProcess) {
      forkedProcess.kill()
    }
  })

  it('works as expected', async function() {
    const communication = new Communication({
      listener(callback) {
        forkedProcess.on('message', callback)
      },
      send(message) {
        forkedProcess.send(message)
      },
    })

    expect(await communication.send('ping', 52)).toBe('pong 52')
    expect(await communication.send('ping2', 53)).toBe('pong2 53')

    try {
      await communication.send('ping3', 54)
      expect(false).toBe(true)
    } catch (error) {
      expect(error.message).toBe('pong3 54')
    }

    try {
      await communication.send('ping4', 55)
      expect(false).toBe(true)
    } catch (error) {
      expect(error.message).toBe('pong4 55')
    }
  })
})


lib/main.js
'use strict';
'use babel';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _sbEventKit = require('sb-event-kit');

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Communication = function () {
  function Communication(debug) {
    var _this = this;

    _classCallCheck(this, Communication);

    this.debug = Boolean(debug);
    this.emitter = new _sbEventKit.Emitter();
    this.subscriptions = new _sbEventKit.CompositeDisposable();

    this.emitter.on('request', function (message) {
      message.response = null;

      var response = new Promise(function (resolve) {
        _this.emitter.emit('request:' + message.name, message.data, message);
        resolve(message.response);
      });

      response.then(function (response) {
        message.response = response;
        return true;
      }, function (val) {
        if (val instanceof Error) {
          var error = { __sb_is_error: true };
          Object.getOwnPropertyNames(val).forEach(function (key) {
            error[key] = val[key];
          });
          message.response = error;
        } else message.response = val;
        return false;
      }).then(function (status) {
        _this.emitter.emit('send', {
          id: message.id, sb_communication: true, status: status, type: 'response', data: message.response
        });
      });
    });
    this.emitter.on('response', function (message) {
      message.data = message.data && message.data.__sb_is_error ? Communication.createError(message.data) : message.data;
      _this.emitter.emit('job:' + message.id, message);
    });
  }

  _createClass(Communication, [{
    key: 'parseMessage',
    value: function parseMessage(messageGiven) {
      var message = void 0;
      try {
        message = typeof messageGiven === 'string' ? JSON.parse(messageGiven) : messageGiven;
      } catch (_) {
        throw new Error('Error decoding response');
      }
      if (!message.sb_communication) {
        // Ignore unknown messages
        return;
      }
      if (this.debug) {
        console.debug(message);
      }
      this.emitter.emit(message.type, message);
    }
  }, {
    key: 'request',
    value: function request(name) {
      var _this2 = this;

      var data = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

      return new Promise(function (resolve, reject) {
        var id = Communication.randomId();
        var disposable = _this2.emitter.on('job:' + id, function (result) {
          disposable.dispose();
          if (result.status) {
            resolve(result.data);
          } else reject(result.data);
        });
        _this2.emitter.emit('send', {
          id: id, sb_communication: true, name: name, type: 'request', data: data
        });
      });
    }
  }, {
    key: 'onRequest',
    value: function onRequest(name, callback) {
      return this.emitter.on('request:' + name, callback);
    }
  }, {
    key: 'onShouldSend',
    value: function onShouldSend(callback) {
      return this.emitter.on('send', callback);
    }
  }, {
    key: 'dispose',
    value: function dispose() {
      this.subscriptions.dispose();
      this.emitter.dispose();
    }
  }], [{
    key: 'randomId',
    value: function randomId() {
      return (Math.random().toString(36) + '00000000000000000').slice(2, 7 + 2);
    }
  }, {
    key: 'createError',
    value: function createError(data) {
      var error = new Error();
      for (var key in data) {
        if (key !== '__sb_is_error') {
          error[key] = data[key];
        }
      }
      return error;
    }
  }]);

  return Communication;
}();

exports.default = Communication;

src/index.js
// @flow

import invariant from 'assert'

type Handle = {|
  send(payload: Object): void,
  listener(callback: (message: Object) => void): void,
|}

class Communication {
  alive: boolean
  handle: Handle
  handlers: { [string]: Function } = {}
  outgoing: { [string]: { resolve: Function, reject: Function } } = {}
  constructor(handle: Handle) {
    invariant(typeof handle === 'object' && handle, 'handle must be a valid object')
    invariant(typeof handle.send === 'function', 'handle.send must be a valid function')
    invariant(typeof handle.listener === 'function', 'handle.listener must be a valid function')

    this.alive = true
    this.handle = handle

    this.handle.listener(message => {
      if (!this.alive || !message.__sb_communication) {
        return
      }
      this.__handleMessage(message)
    })
  }
  __getId() {
    let id
    do {
      id = Math.random()
        .toString(36)
        .substring(7)
    } while (this.outgoing[id])
    return id
  }
  __handleMessage(message: Object) {
    if (message.type === 'request') {
      const response = {
        __sb_id: message.__sb_id,
        __sb_communication: true,
        successful: true,
        payload: undefined,
        type: 'response',
      }
      const handler = this.handlers[message.event]
      if (!handler) {
        if (this.alive) {
          this.handle.send(message)
        }
      } else {
        const result = new Promise(function(resolve) {
          resolve(handler(message.payload))
        })
        result
          .then(payload => {
            response.payload = payload
          })
          .catch(error => {
            response.successful = false
            response.payload = error
          })
          .then(() => {
            if (this.alive) {
              if (response.payload && typeof response.payload === 'object') {
                const masked = {}
                Object.getOwnPropertyNames(response.payload).forEach(function(key) {
                  masked[key] = response.payload[key]
                })
                response.payload = masked
              }
              this.handle.send(response)
            }
          })
      }
    } else if (message.type === 'response') {
      const outgoing = this.outgoing[message.__sb_id]
      if (outgoing) {
        this.outgoing[message.__sb_id] = null
        outgoing[message.successful ? 'resolve' : 'reject'](message.payload)
      }
    }
  }
  on(event: string, callback: Function, force: boolean = false) {
    invariant(typeof event === 'string' && event, 'event must be a valid string')
    invariant(typeof callback === 'function', 'callback must be a valid function')

    if (!force && this.handlers[event]) {
      throw new Error(`Event '${event}' has a listener setup already`)
    }
    this.handlers[event] = callback
  }
  send(event: string, payload: any): Promise<any> {
    return new Promise((resolve, reject) => {
      if (!this.alive) {
        reject(new Error('Communication bridge is no longer active'))
        return
      }

      const message = {
        __sb_id: this.__getId(),
        __sb_communication: true,
        event,
        payload,
        type: 'request',
      }
      this.outgoing[message.__sb_id] = { resolve, reject }
      this.handle.send(message)
    })
  }
  dispose() {
    this.alive = false
  }
}

module.exports = Communication


sb-communication.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.EventKit = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
'use strict';
'use babel';

Object.defineProperty(exports, "__esModule", {
  value: true
});
class CompositeDisposable {
  constructor() {
    this.disposed = false;
    this.disposables = new Set(arguments);
  }
  dispose() {
    if (!this.disposed) {
      this.disposed = true;
      this.disposables.forEach(function (item) {
        item.dispose();
      });
      this.disposables = null;
    }
  }
  add() {
    var _this = this;

    if (!this.disposed) {
      Array.prototype.forEach.call(arguments, function (item) {
        return _this.disposables.add(item);
      });
    }
  }
  remove() {
    var _this2 = this;

    if (!this.disposed) {
      Array.prototype.forEach.call(arguments, function (item) {
        return _this2.disposables.delete(item);
      });
    }
  }
  clear() {
    if (!this.disposed) {
      this.disposables.clear();
    }
  }
}
exports.CompositeDisposable = CompositeDisposable;
},{}],2:[function(require,module,exports){
'use strict';
'use babel';

Object.defineProperty(exports, "__esModule", {
  value: true
});
class Disposable {
  constructor(callback) {
    this.disposed = false;
    this.callback = callback;
  }
  dispose() {
    if (!this.disposed) {
      this.disposed = true;
      if (typeof this.callback === 'function') {
        this.callback();
      }
      this.callback = null;
    }
  }
}
exports.Disposable = Disposable;
},{}],3:[function(require,module,exports){
'use strict';
'use babel';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Emitter = undefined;

var _disposable = require('./disposable');

class Emitter {
  constructor() {
    this.disposed = false;
    this.handlers = {};
  }
  dispose() {
    this.disposed = true;
    this.handlers = null;
  }
  on(eventName, handler) {
    var _this = this;

    if (this.disposed) {
      throw new Error('Emitter has been disposed');
    }
    if (typeof handler !== 'function') {
      throw new Error('Event handler must be a function');
    }
    if (typeof this.handlers[eventName] === 'undefined') {
      this.handlers[eventName] = [handler];
    } else {
      this.handlers[eventName].push(handler);
    }
    return new _disposable.Disposable(function () {
      _this.off(eventName, handler);
    });
  }
  off(eventName, handler) {
    if (this.disposed || typeof this.handlers[eventName] === 'undefined') {
      return;
    }
    const index = this.handlers[eventName].indexOf(handler);
    if (index !== -1) {
      this.handlers[eventName].splice(index, 1);
    }
  }
  clear() {
    this.handlers = [];
  }
  emit(eventName) {
    for (var _len = arguments.length, params = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
      params[_key - 1] = arguments[_key];
    }

    if (this.disposed || typeof this.handlers[eventName] === 'undefined') {
      return;
    }
    const paramsLength = params.length;
    this.handlers[eventName].forEach(function (callback) {
      if (paramsLength === 1) {
        callback(params[0]);
      } else if (paramsLength === 2) {
        callback(params[0], params[1]);
      } else {
        callback.apply(undefined, params);
      }
    });
  }
}
exports.Emitter = Emitter;

},{"./disposable":2}],4:[function(require,module,exports){
'use strict';
'use babel';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _disposable = require('./disposable');

Object.defineProperty(exports, 'Disposable', {
  enumerable: true,
  get: function () {
    return _disposable.Disposable;
  }
});

var _compositeDisposable = require('./composite-disposable');

Object.defineProperty(exports, 'CompositeDisposable', {
  enumerable: true,
  get: function () {
    return _compositeDisposable.CompositeDisposable;
  }
});

var _emitter = require('./emitter');

Object.defineProperty(exports, 'Emitter', {
  enumerable: true,
  get: function () {
    return _emitter.Emitter;
  }
});
},{"./composite-disposable":1,"./disposable":2,"./emitter":3}],5:[function(require,module,exports){
'use strict';
'use babel';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _sbEventKit = require('sb-event-kit');

class Communication {
  constructor(debug) {
    var _this = this;

    this.debug = Boolean(debug);
    this.emitter = new _sbEventKit.Emitter();
    this.subscriptions = new _sbEventKit.CompositeDisposable();

    this.emitter.on('request', function (message) {
      message.response = null;

      const response = new Promise(function (resolve) {
        _this.emitter.emit(`request:${ message.name }`, message.data, message);
        resolve();
      });

      response.then(function () {
        return true;
      }, function (val) {
        if (val instanceof Error) {
          const error = { __sb_is_error: true };
          Object.getOwnPropertyNames(val).forEach(function (key) {
            error[key] = val[key];
          });
          message.response = error;
        } else message.response = val;
        return false;
      }).then(function (status) {
        _this.emitter.emit('send', {
          id: message.id, sb_communication: true, status: status, type: 'response', data: message.response
        });
      });
    });
    this.emitter.on('response', function (message) {
      message.data = message.data && message.data.__sb_is_error ? Communication.createError(message.data) : message.data;
      _this.emitter.emit(`job:${ message.id }`, message);
    });
  }
  parseMessage(messageGiven) {
    let message;
    try {
      message = typeof messageGiven === 'string' ? JSON.parse(messageGiven) : messageGiven;
    } catch (_) {
      throw new Error('Error decoding response');
    }
    if (!message.sb_communication) {
      // Ignore unknown messages
      return;
    }
    if (this.debug) {
      console.debug(message);
    }
    this.emitter.emit(message.type, message);
  }
  request(name) {
    var _this2 = this;

    let data = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

    return new Promise(function (resolve, reject) {
      const id = Communication.randomId();
      const disposable = _this2.emitter.on(`job:${ id }`, function (result) {
        disposable.dispose();
        if (result.status) {
          resolve(result.data);
        } else reject(result.data);
      });
      _this2.emitter.emit('send', {
        id: id, sb_communication: true, name: name, type: 'request', data: data
      });
    });
  }
  onRequest(name, callback) {
    return this.emitter.on(`request:${ name }`, callback);
  }
  onShouldSend(callback) {
    return this.emitter.on('send', callback);
  }
  dispose() {
    this.subscriptions.dispose();
    this.emitter.dispose();
  }
  static randomId() {
    return (Math.random().toString(36) + '00000000000000000').slice(2, 7 + 2);
  }
  static createError(data) {
    const error = new Error();
    for (const key in data) {
      if (key !== '__sb_is_error') {
        error[key] = data[key];
      }
    }
    return error;
  }
}
exports.default = Communication;
},{"sb-event-kit":4}]},{},[5])(5)
});